src/charm.py

```diff
@@ -127,13 +128,8 @@ def __init__(self, *args):
         self.karma_provider = KarmaProvider(self, "karma-dashboard")
         self.remote_configuration = RemoteConfigurationRequirer(self)
 
-        api = ServicePort(self._ports.api, name=f"{self.app.name}")
-        ha = ServicePort(self._ports.ha, name=f"{self.app.name}-ha")
-        self.service_patcher = KubernetesServicePatch(
-            charm=self,
-            service_type="ClusterIP",
-            ports=[api, ha],
-        )
+        self.set_ports()
```

Comment:
Shouldn't we catch a possibly `ModelError` exception raised by `open_port` or `close_port`?

Reply:
We can let the charm goes into `ErrorStatus`, there is no problem with that.

My question is:  Before that, shouldn't we "unset" the ports that were setted?

Perhaps I am completely wrong, but seems like `set_ports()` should be a transactional action.

Reply:
> shouldn't we "unset" the ports that were setted?

We could, but I imagine in this case we'd rather have "degraded mode" than nonoperational.
It's also "corrected" every hook.